(WIP) Web Usernames MVP

SabreCat opened the pull request:
**NOT FOR PRODUCTION AT THIS TIME** - This is meant to go live when we are ready to require usernames for all users.

Fixes #10697 
Fixes https://github.com/HabitRPG/habitica-private/issues/12
Fixes https://github.com/HabitRPG/habitica-private/issues/17
Fixes https://github.com/HabitRPG/habitica-private/issues/18

## Username Verification Modals

* Adds a modal that appears automatically if the user has not yet verified their username. The modal gives the user the opportunity to change their display name and/or username and cannot be dismissed except by providing valid input for both fields.
* Adjusts z-indexing of modals so they don't fall behind the "Resting in the Inn" banner. Again. Hopefully.
* Changes the first page of the Justin onboarding flow to give the user a chance to review and update their display name and username.

## Name and Email Verification

* Adds an API route `verify-display-name`, based on `verify-username`, so we can check display names for slurs and length inline.
* Cleans out a few unused name-related strings.
* Fixes an error during email validation in front page login/register flows.
* Replaces in-component regex on emails with the `isEmail` method from `validator`.

## Chat and Private Message Updates

* Cleans up a bunch of styling in Guild/Tavern/Party chat to better fit designs.
* Adds username to the chat message model.
* Shows @username at the top of chat messages, when present.
* Shows both display name (fka profile name) and @username in autocomplete searches, and allows searching on either name.
* Limits autocomplete searches to top 5 matches.
* Adds mouseover highlighting to autocomplete options.
* Adds highlight to @mentions in chat messages.
* Shows @username in inbox conversation list, and implements the full intended styling for profile names there.
* Shows @username in party header, including mouseover for members.
* Shows @username in member profile modal.

## Group Invitation Updates

* Allows users to invite others to groups via username, based on @TheHollidayInn's work.
* Adds an API route `GET /members/username/:username` to allow lookup of public user data via username. It returns not-found if the target username is unverified.

TheHollidayInn committed:
Added username invite

TheHollidayInn committed:
Merged in develop

SabreCat committed:
feat(usernames): modal to force verification

SabreCat committed:
fix(test): string is miscapitalized :(

SabreCat committed:
fix(usernames): remove more ratzen fratzen dupe strings

SabreCat committed:
fix(test): copypasta and string trouble

SabreCat committed:
fix(username): Add @ prepend

SabreCat committed:
fix(usernames): add display name verification to site settings
also corrrect some validation logic in force-verify modal

SabreCat committed:
fix(settings): don't center align display name issues

SabreCat committed:
fix(usernames): various
Better modal positioning
Correct text colors for input field
Don't show "already taken" if username has other errors

SabreCat committed:
WIP(usernames): display in chat areas

SabreCat committed:
fix(model): don't break if auth.local undefined

SabreCat committed:
feat(usernames): Veteran Pet ladder award for affected users

SabreCat committed:
feat(usernames): vet pet announcement

SabreCat committed:
Merge branch 'sabrecat/force-username-modal' into sabrecat/usernames-master

SabreCat committed:
Merge branch 'sabrecat/username-display' into sabrecat/usernames-master

SabreCat committed:
Merge branch 'sabrecat/veteran-verified' into sabrecat/usernames-master

SabreCat committed:
chore(sprites): compile

SabreCat committed:
fix(chat): better likes and display name font sizing

SabreCat committed:
Merge branch 'sabrecat/username-display' into sabrecat/usernames-master

SabreCat committed:
fix(usernames): let verify modal grow to content

SabreCat committed:
Merge branch 'sabrecat/force-username-modal' into sabrecat/usernames-master

SabreCat committed:
fix(chat): padding adjustment and tooltip placement

SabreCat committed:
Merge branch 'sabrecat/username-display' into sabrecat/usernames-master

SabreCat committed:
fix(chat): match hyphen in @name regex

SabreCat committed:
Merge branch 'sabrecat/username-display' into sabrecat/usernames-master

SabreCat committed:
feat(usernames): show in party header, profiles, inbox convo list

SabreCat committed:
Merge branch 'sabrecat/username-display' into sabrecat/usernames-master

SabreCat committed:
fix(usernames): various
z-index modals above Resting banner
force reload after verify username
add missing e-mail validation on frontpage
let Yesterdaily modal float behind username modal

SabreCat committed:
Merge branch 'sabrecat/force-username-modal' into sabrecat/usernames-master

SabreCat committed:
fix(chat): better @ searching, no chat card borders

SabreCat committed:
Merge branch 'sabrecat/username-display' into sabrecat/usernames-master

SabreCat committed:
fix(chat): no min height on autocomplete dropdown

SabreCat committed:
Merge branch 'sabrecat/username-display' into sabrecat/usernames-master

SabreCat committed:
fix(profile): better username/UUID styling

SabreCat committed:
Merge branch 'sabrecat/username-display' into sabrecat/usernames-master

SabreCat committed:
fix(PR): remove unrelated changes

SabreCat committed:
Merge branch 'develop' into sabrecat/usernames-master

SabreCat committed:
fix(PR): more cleanup

SabreCat committed:
fix(comments): validator for emails, remove prefixes, allow length 1

SabreCat commented:
@paglias Addressed comments via https://github.com/HabitRPG/habitica/pull/10784/commits/5299c8d406ddd7efed355f1a79088f7453a961c0 - now that we're using `validator`, what do you think should be in a mixin? Bundle up the various email/username/etc. validations I guess?

SabreCat committed:
feat(usernames): verification step during Justin intro for social users

paglias commented:
@SabreCat I had some trouble moving the functions to a mixin because there were some slight changes between them but if we manage to it would be great

Just one thing about validator, since it's quite big can we load just the necessary functions? So `import validator/lib/isEmail`

SabreCat committed:
Merge branch 'release' into sabrecat/usernames-master

SabreCat committed:
fix(chat): replace autocomplete at @

SabreCat committed:
refactor(auth): only import needed validator module

SabreCat committed:
fix(usernames): don't supply username in public fields if unverified

SabreCat committed:
fix(inbox): display correct UN for outbound user

SabreCat committed:
fix(inbox): more UN display fixes

SabreCat committed:
fix(inbox): don't show likes in inbox
Also remove convo list contrib styling for now

SabreCat committed:
Merge branch 'develop' into sabrecat/usernames-master

SabreCat committed:
fix(modal): various

SabreCat committed:
fix(modals): better stack???

SabreCat committed:
fix(modals): maybe got it?!?

SabreCat committed:
fix(modals): button styling

SabreCat committed:
fix(usernames): RYA positioning and contrib tier in PMs

SabreCat committed:
fix(likes): accessibility tweaks

SabreCat committed:
fix(auth): new users should start verified

SabreCat committed:
Merge branch 'release' into develop

SabreCat committed:
fix(usernames): various
Partial fixage for autocomplete @ing
Don't add username to chat message if user is unverified
Fix flying pets
Fix console error about avatar intro

SabreCat committed:
Merge branch 'develop' into sabrecat/usernames-master

SabreCat committed:
Merge branch 'develop' into sabrecat/usernames-master

SabreCat committed:
feat(usernames): invite by username

SabreCat committed:
fix(invites): bogus validation errors

paglias commented on website/server/controllers/api-v3/members.js:
```diff
@@ -119,6 +119,33 @@ api.getMember = {
+    let username = req.params.username.toLowerCase();
+
+    let member = await User
+      .findOne({'auth.local.lowerCaseUsername': username})
```
are lowercase usernames unique though?

SabreCat replied:
oh, blurgh, you're right. I suppose this could be set up to return multiple...

SabreCat committed:
Merge branch 'develop' into sabrecat/usernames-master

SabreCat committed:
fix(usernames): various
Disappearing input fields
Text replacement in @mentions
UUID visibility in profiles
Purple dot for @mentioned usernames
TypeError preventing RYA
Group Plan member list error

SabreCat committed:
Merge branch 'develop' into sabrecat/usernames-master

SabreCat committed:
fix(invites): show errors inline

SabreCat committed:
fix(usernames): exclude unverified users during query

SabreCat committed:
fix(usernames): filter @ on server side for username lookup

SabreCat committed:
fix(usernames): address failing tests

SabreCat committed:
fix(invites): more responsive validation

SabreCat committed:
fix(tests): correct expects

SabreCat committed:
fix(tests): linting & more expects
Also one more tweak for invite validation responsiveness
